drivers/spi/spi_ite_it8xxx2.c

```diff
@@ -0,0 +1,496 @@
+			cmdq_header_txbuf = txbfptr;
+		}
+
+	}
+
+	/* Assign Addresses for Header and the Read Buffer */
+	spi_it8xxx2_cmdq_header_setup(dev,
+		wr_cmd_length,
+		command1,
+		data_length,
+		(uint32_t)data_buf_addr,
+		check_bit_mask,
+		check_bit_value,
+		cmdq_header_txbuf);
+
+	/* Assign Addresses for Header and the Read Buffer */
+	spi_it8xxx2_cmdq_buffer_mapping(dev,
+					cfg->inst_no,
+					data->spi_cmdq_header_addr,
+					data->spi_cmdq_read_buff_addr);
+
+	chip_block_idle();
+	spi_regs->SPICTRL5 |= (cfg->inst_no) ? (CH1SELCMDQ) : (CH0SELCMDQ);
+	spi_regs->SPICTRL5 |= CMDQMODE;
+
+	int ret = k_sem_take(&it8xxx2_sem, K_FOREVER);
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = -ETIMEDOUT;
+	}
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	return err;
+
+}
+
+static int it8xxx2_transceive(const struct device *dev,
+			    const struct spi_config *config,
+			    const struct spi_buf_set *tx_bufs,
+			    const struct spi_buf_set *rx_bufs)
+{
+	size_t nbrx = 0;
+	size_t nbtx = 0;
+
+	const struct spi_buf *ptx = tx_bufs->buffers;
+	const struct spi_buf *prx = rx_bufs->buffers;
+
+	int err = 0;
+	int ret = 0;
+
+	ret = k_mutex_lock(&it8xxx2_mutex, K_FOREVER);
+
+	if (tx_bufs != NULL) {
+		nbtx = tx_bufs->count;
+	}
+
+	while (nbtx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	if (rx_bufs != NULL) {
+		nbrx = rx_bufs->count;
+	}
+
+	while (nbrx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	k_mutex_unlock(&it8xxx2_mutex);
+
+	return err;
+
+}
+
+static int it8xxx2_transceive_sync(const struct device *dev,
+				 const struct spi_config *config,
+				 const struct spi_buf_set *tx_bufs,
+				 const struct spi_buf_set *rx_bufs)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int it8xxx2_transceive_async(const struct device *dev,
+				  const struct spi_config *config,
+				  const struct spi_buf_set *tx_bufs,
+				  const struct spi_buf_set *rx_bufs,
+				  struct k_poll_signal *async)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+#endif
+
+static int it8xxx2_release(const struct device *dev,
+			 const struct spi_config *config)
+{
+	k_sem_give(&it8xxx2_sem);
+	return 0;
+}
+
+static void spi_it8xxx2_isr(const struct device *dev)
+{
+	const struct spi_it8xxx2_config *cfg = dev->config;
+	struct spi_it8xxx2_regs *const spi_regs =
+		(struct spi_it8xxx2_regs *)cfg->base_addr;
+
+	irq_disable(cfg->irq_no);
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	ite_intc_isr_clear(cfg->irq_no);
+	spi_regs->SPICTRL5 &= ~(CH1SELCMDQ | CH0SELCMDQ);
+	k_sem_give(&it8xxx2_sem);
+	chip_permit_idle();
+}
```

Comment:
This ISR doesn't appear to support configurations where both SPI controllers are enabled.  The 2nd SPI controller initialized will never generate or process any interrupts.

You likely need shared interrupt support as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/49427.

The alternative here would be to use the `DT_INST_FOREACH_STATUS_OKAY()` and `DEVICE_DT_INST_GET()` to lookup the the `struct device` for all SPI controllers, and check if there is pending interrupt on each controller.  I think you would also need an semaphore for each SPI controller instance instead of shared global one.

Reply:
Dear Keith,

Sorry for delayed response coz we were in a long vacation of Lunar New Year.
Yes, I think that you're absolutely right, the interrupt connection only works for one instance.

For the instance 0 and 1 share the same interrupt number, I'm wondering if it's still possible to tell if there is pending interrupt on each controller. 

I think that the #49427 which you mentioned matches the requirement we'd need here.
I'll back to this after some internal discussions.

Truly appreciative of your guidance & helps.

Warmest regards,
BJ

Reply:
See this [example](https://github.com/zephyrproject-rtos/zephyr/blob/a2f203dd1a7a5110a3339da81e72da024be05ab5/drivers/adc/adc_gd32.c#L419) for how to handle a shared interrupt by two driver instances:

To adapt this to the ITE SPI driver, you would read the INTSTS register to determine if an interrupt is pending:

```
spi_it8xxx2_instance_isr(dev)
{
    /* Clear interrupts, give semaphore, etc */
}

#define HANDLE_SHARED_IRQ(n, active_irq)	
	static const struct device *const dev_##n = DEVICE_DT_INST_GET(n);
	const struct spi_it8xxx2_regs *spi_regs_##n = dev_##n->config->base_addr;;
	if (spi_regs_##->INTSTS) {
		spi_it8xxx2_instance_isr(dev_##);    
	}
void spi_it8xxx2_shared_isr(const struct device *dev)
{
	ARG_UNUSED(dev);
	DT_INST_FOREACH_STATUS_OKAY(HANDLE_SHARED_IRQ);
}
```

Reply:
Thank you, sir. We also found that our chip can only support spi0 according to the current pin numbers, so specific chip with more pins can support second instance but so far we don't have it. So the second instance has been removed in dtsi files. Though we keep the init style like multiple instance but it's just for the scalability in the future.